### ISSUE TITLE
View op test case invalid syntax

### ISSUE THREAD

**Issue by bondhugula** [tensorflow/mlir#286]
The syntax here appears to be messed up. 
https://github.com/tensorflow/mlir/blob/b7e6270e6a06fb67b5215e1647774143abaf506c/test/Dialect/Linalg/loops.mlir#L19

From ops.td, the view op takes the offset first followed by sizes (reproduced below), while the test case above passes sizes followed by offset. Not sure why the test case IR even verifies. 
``` 
// ViewOp with dynamic offset and one dynamic size.
    %2 = view %0[%offset_1024][%size0]
      : memref<2048xi8> to memref<?x4xf32, (d0, d1)[s0] -> (d0 * 4 + d1 + s0)>
```

**Comment by ftynse** [comment]
The parser seems to accept a list of offsets https://github.com/tensorflow/mlir/blob/master/lib/Dialect/StandardOps/Ops.cpp#L2332 and the verifier does not catch it.

**Comment by bondhugula** [comment]
> The parser seems to accept a list of offsets https://github.com/tensorflow/mlir/blob/master/lib/Dialect/StandardOps/Ops.cpp#L2332 and the verifier does not catch it.

The number of 'sizes' is also one less!

**Comment by ftynse** [comment]
This is deeper than I thought, in several ways. I have a pending fix for the ordering problem, and there were incorrect assumptions in other places.

**Comment by ftynse** [comment]
> The number of 'sizes' is also one less!

I don't follow this. The resulting type is `memref<?x?xf32, ...>`, which has two dynamic sizes and we pass two values as sizes.

**Comment by bondhugula** [comment]
> This is deeper than I thought, in several ways. I have a pending fix for the ordering problem, and there were incorrect assumptions in other places.

Just FYI - my PR #253 fixes the loops.mlir file itself - it had been approved long ago (but not committed), and I discovered this bug when I tried to resolve conflicts there.

**Comment by ftynse** [comment]
Thanks! There are more tests that need updating to fix this though. Your PR is currently blocked on internal integration due to test breakage, I pinged the relevant person.

**Comment by bondhugula** [comment]
> > The number of 'sizes' is also one less!
> 
> I don't follow this. The resulting type is `memref<?x?xf32, ...>`, which has two dynamic sizes and we pass two values as sizes.

`%A = view %arg0[%M, %K][%c0] :` <-- only %c0 is being passed in as size.

**Comment by ftynse** [comment]
Ah ok, I see. My interpretation is that the order of []-blocks is swapped. `[%M, %K]` are sizes `[%c0]` is the offset.

**Comment by nicolasvasilache** [comment]
This was the old syntax, it was changed to be consistent with subview but
the parser/verifier were not hardened and tests not updated.
@Oleksandr Zinenko <ftynse@gmail.com> has a fix landing, thanks for
reporting!

**Comment by ftynse** [comment]
Fixed in 721a07c
